src/controllers/colormapper.cpp

```diff
@@ -0,0 +1,59 @@
+#include "controllers/colormapper.h"
+
+#include <cmath>
+#include <cstdint>
+
+#include "util/debug.h"
+
+namespace {
+double colorDistance(QRgb a, QRgb b) {
+    // This algorithm calculates the distance between two colors. In
+    // contrast to the L2 norm, this also tries take the human perception
+    // of colors into account. More accurate algorithms like the CIELAB2000
+    // Delta-E rely on sophisticated color space conversions and need a lot
+    // of costly computations. In contrast, this is a low-cost
+    // approximation and should be sufficently accurate.
+    // More details: https://www.compuphase.com/cmetric.htm
+    long mean_red = (static_cast<long>(qRed(a)) + static_cast<long>(qRed(b))) / 2;
+    long delta_red = static_cast<long>(qRed(a)) - static_cast<long>(qRed(b));
+    long delta_green = static_cast<long>(qGreen(a)) - static_cast<long>(qGreen(b));
+    long delta_blue = static_cast<long>(qBlue(a)) - static_cast<long>(qBlue(b));
+    return sqrt(
+            (((512 + mean_red) * delta_red * delta_red) >> 8) +
+            (4 * delta_green * delta_green) +
+            (((767 - mean_red) * delta_blue * delta_blue) >> 8));
+}
+} // namespace
+
+QPair<QRgb, QVariant> ColorMapper::getNearestColor(QRgb desiredColor) {
+    // If desired color is already in cache, use cache entry
+    QMap<QRgb, QRgb>::const_iterator i = m_cache.find(desiredColor);
+    QMap<QRgb, QVariant>::const_iterator j;
+    if (i != m_cache.constEnd()) {
+        j = m_availableColors.find(i.value());
+        DEBUG_ASSERT(j != m_availableColors.constEnd());
+        qDebug() << "ColorMapper cache hit for" << desiredColor << ":"
+                 << "Color =" << j.key() << ","
+                 << "Value =" << j.value();
+        return QPair<QRgb, QVariant>(j.key(), j.value());
+    }
+
+    // Color is not cached
+    QMap<QRgb, QVariant>::const_iterator nearestColorIterator;
+    double nearestColorDistance = qInf();
+    for (j = m_availableColors.constBegin(); j != m_availableColors.constEnd(); j++) {
+        QRgb availableColor = j.key();
+        double distance = colorDistance(desiredColor, availableColor);
+        if (distance < nearestColorDistance) {
+            nearestColorDistance = distance;
+            nearestColorIterator = j;
+        }
+    }
+
+    DEBUG_ASSERT(nearestColorDistance < qInf());
```

Comment:
is the idea that this should just never happen? is there a safe default we can return on error instead of all these asserts?

Reply:
yeah, this should only happen if `m_availableColors` is empty which should not happen until someone doesn't pass any color to `ColorMapper` (which is invalid).

Reply:
on second thought I'll leave it in just in case someone messes with the code inside that function and breaks something ;-)

Reply:
that case should *definitely* be covered by a unit test, and this class is small and self-contained and so it should have a testing file.  You can add a case where it passes various bad options and see if you can break it.  Adding asserts in every place where someone could possibly mess up the code would just mean adding asserts for essentially every invariant at every stage of logic.